### ISSUE TITLE
Segfault when clicking a certain lore.kernel.org link in a certain gregkh tweet

### ISSUE THREAD

**Issue by lucaswerkmeister** [IBBoard/cawbird#374]
**Describe the bug**
If I click the link in [this tweet](https://twitter.com/gregkh/status/1384785747874656257) in Cawbird, it segfaults. The issue seems to be perfectly reproducible and happens no matter whether the tweet appears in the timeline, focused in tweet details view, or as one of the ancestor tweets in the tweet details view (e.g. when viewing [this reply](https://twitter.com/gregkh/status/1384857799058984963)).

**To Reproduce**
1. Open [this tweet](https://twitter.com/gregkh/status/1384785747874656257), e.g. by finding it on gregkh’s profile
2. Click the link
3. Cawbird crashes

**Expected behavior**
Open the link in the browser.

**System details:**
 - OS: Arch Linux
 - Version: 1.3.2.r87.g5940a5f8-1 (5940a5f8)
 - Package type: [cawbird-git](https://aur.archlinux.org/packages/cawbird-git)

**Additional context**
Another lore.kernel.org link in a different gregkh tweet (the reply mentioned above) is fine. Right-clicking the “bad” link and copying its URL also works without issue.

I have not yet tested if the same link in a different tweet triggers the issue; I’ll do that later (and also try latest master, since I’m probably a bit behind). (TODO: Running `G_MESSAGES_DEBUG=cawbird cawbird` can give useful debug output.)

<details>
<summary>Backtrace</summary>

```
#0  0x00007f0a11c611bc in g_utf8_validate () at /usr/lib/libglib-2.0.so.0
#1  0x00007f0a113b1073 in  () at /usr/lib/libgdk-3.so.0
#2  0x00007f0a11045f31 in  () at /usr/lib/libgio-2.0.so.0
#3  0x00007f0a11046603 in  () at /usr/lib/libgio-2.0.so.0
#4  0x00007f0a1102a3d4 in  () at /usr/lib/libgio-2.0.so.0
#5  0x00007f0a1102e7b9 in  () at /usr/lib/libgio-2.0.so.0
#6  0x00007f0a11088753 in  () at /usr/lib/libgio-2.0.so.0
#7  0x00007f0a1102a3d4 in  () at /usr/lib/libgio-2.0.so.0
#8  0x00007f0a1102a409 in  () at /usr/lib/libgio-2.0.so.0
#9  0x00007f0a11c31f30 in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#10 0x00007f0a11c85b59 in  () at /usr/lib/libglib-2.0.so.0
#11 0x00007f0a11c2f781 in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
#12 0x00007f0a1105722e in g_application_run () at /usr/lib/libgio-2.0.so.0
#13 0x000055fbc57e614c in _vala_main (args=0x7ffd5c9876e8, args_length1=2) at ../src/main.vala:35
#14 0x000055fbc57e6294 in main (argc=2, argv=0x7ffd5c9876e8) at ../src/main.vala:18
```

</details>

**Comment by IBBoard** [comment]
It's not being helpful with that backtrace. But I suspect it'll be something weird to do with the @ or something.

**Comment by lucaswerkmeister** [comment]
Issue still reproducible on master (d300abf290a9b78d856362fb38af3269aab7727d), not much useful debug output:

```
(cawbird:24742): cawbird-DEBUG: 20:46:28.050: MainWindow.vala:60: Focus widget now: TweetListEntry 0x562e9feeda20                                                                                                                                                                                     
(cawbird:24742): cawbird-DEBUG: 20:46:28.051: TweetUtils.vala:890: Activating 'https://lore.kernel.org/linux-nfs/YH%2FfM%2FTsbmcZzwnX@kroah.com/'                                                                                                                                                     
Segmentation fault (core dumped)                                                                                                                                                                                                                                                                      
```

Trying to `gdb` my way through it now.

**Comment by IBBoard** [comment]
Too many of the useful fields seem to be "optimized out" in GDB. So I've taken the other approach: https://twitter.com/IBBTwtr/status/1384944016668741634

It's the %2F (URL encoded slash). The third link is fine but the first two crash.

Now to work out what we're doing wrong with URL encoded slashes.

Presumably this is related to fixing #305 😄

**Comment by IBBoard** [comment]
I'm slightly suspicious that you found a GTK bug!

The debugging shows that it hits out `activate_link` function. We do special things with @-handle, #hashtag and https://twitter.com URLs, but then let it fall back to [default link clicking behaviour](https://developer.gnome.org/gtk3/stable/GtkLabel.html#GtkLabel-activate-link), which is [gtk_show_uri_on_window()](https://developer.gnome.org/gtk3/stable/gtk3-Filesystem-utilities.html#gtk-show-uri-on-window).

I've just added this one line with a fixed and correct URL as a quick test:

```
return Gtk.show_uri_on_window(window, "https://lore.kernel.org/linux-nfs/YH%2FfM%2FTsbmcZzwnX@kroah.com/", 100);
````

I made every link open this. And it crashed Cawbird with the same error. So it's not a parsing error. It's literally just that `Gtk.show_uri_on_window` doesn't like that URL. If I manually swap `%2F` for `/` then it works fine.

Looks like it's time for a quick bit of digging in GTK/GLib's source code!
